src/invite-api.js

```diff
@@ -1,236 +1,440 @@
 // @ts-check
 import { TypedEmitter } from 'tiny-typed-emitter'
 import { InviteResponse_Decision } from './generated/rpc.js'
-import { projectKeyToId, projectKeyToPublicId } from './utils.js'
+import { assert, keyToId } from './utils.js'
+import timingSafeEqual from './lib/timing-safe-equal.js'
+import { promiseWithResolvers } from './ponyfills.js'
 
 /**
- * @typedef {Object} InviteApiEvents
+ * Internally, we typically use the `Invite` type from the protobuf. We also use
+ * an external type for public consumers.
  *
- * @property {(info: { projectId: string, projectName?: string, peerId: string, roleName: string, roleDescription: string | undefined, invitorName: string }) => void} invite-received
+ * @internal
+ * @typedef {import('./generated/rpc.js').Invite} InviteInternal
+ */
+
+/**
+ * @internal
+ * @typedef {import('./generated/rpc.js').ProjectJoinDetails} ProjectJoinDetails
  */
 
 /**
- * @template K
- * @template V
- * @extends {Map<K, Set<V>>}
+ * @typedef {object} Invite
+ * @prop {string} inviteId
+ * @prop {string} projectName
+ * @prop {string} [roleName]
+ * @prop {string} [roleDescription]
+ * @prop {string} [invitorName]
  */
-class MapOfSets extends Map {
-  /** @param {K} key */
-  get(key) {
-    const existing = super.get(key)
-    if (existing) return existing
-    /** @type {Set<V>} */
-    const set = new Set()
-    super.set(key, set)
-    return set
+
+/**
+ * Manage pending invite state.
+ */
+class PendingInvites {
+  /**
+   * @internal
+   * @typedef {object} PendingInvite
+   * @prop {string} peerId
+   * @prop {InviteInternal} invite
+   * @prop {boolean} isAccepting
+   */
+
+  /** @type {Map<string, PendingInvite>} */
+  #byInviteId = new Map()
+
+  /** @type {Map<string, PendingInvite[]>} */
+  #byProjectPublicId = new Map()
+
+  /**
+   * @returns {Iterable<PendingInvite>} the pending invites, in insertion order
+   */
+  invites() {
+    return this.#byInviteId.values()
+  }
+
+  /**
+   * @param {PendingInvite} pendingInvite
+   * @throws if adding a duplicate invite ID
+   * @returns {void}
+   */
+  add(pendingInvite) {
+    const { invite } = pendingInvite
+    const { inviteId, projectPublicId } = invite
+
+    const inviteIdString = keyToId(inviteId)
```

Comment:
I named this so _of course_ it seems reasonable to me...but `inviteKey` implies cryptographic properties to me. I think `inviteId` is the clearest, but I'm happy to be swayed.

Reply:
Yeah, I agree. We had a convention earlier on to use `key` for buffers, and `id` for their string representation, but we didn't stick to it, and I think the implication of cryptographic properties is confusing. It would be good to adopt a naming convention however that does clearly identify is something is a buffer or a string representation of the same thing.